Add support for VK_EXT_debug_utils to framework, including a new sample and tutorial

SaschaWillems opened the pull request:
## Description

This PR adds:

### Support for the newer VK_EXT_debug_utils extension
The framework now checks if VK_EXT_debug_utils is present, and if so uses it over the deprecated VK_EXT_debug_report. This includes a new debug messenger callback that'll output validation layer messages similar to the old debug callback.

### A new example demonstrating the use of the VK_EXT_debug_utils extension
A new extension example was added that shows how to use the new functions provided by this extension to add labels and object names to a Vulkan application to be displayed in a debugger.
**Note:** In order to test the functionality of this sample it needs to be run from a debugger like RenderDoc.

### A tutorial on the VK_EXT_debug_utils extension
The above sample comes with a new tutorial that talks about the features of the new extension, and shows how to use it along with RenderDoc.

Fixes #47 

## General Checklist:
- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

## Sample Checklist
- [x] I have tested the sample on at least one compliant Vulkan implementation
- n/a] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- n/a Any dependent assets have been merged and published in downstream modules

The sample has been tested on Windows with Vulkan 1.2 (using the new extension) and on Android with Vulkan 1.0 (testing fallback to the old extension).

SaschaWillems committed:
Add support for validation with VK_EXT_Debug_Utils

Will fall back to the old VK_EXT_Debug_Report extensions if not supported

SaschaWillems committed:
Add new extensions sample for VK_EXT_Debug_Utils

SaschaWillems committed:
Destroy debug utils callback

SaschaWillems committed:
Only use debug utils functions of extension is available

SaschaWillems committed:
Add name to skysphere texture image

Add queue label

SaschaWillems committed:
Added tutorial for using the debug utilities extension with a debugger like RenderDoc

SaschaWillems committed:
Fix clang format and copyright headers

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

gary-sweet reviewed (dismissed):
This doesn't work in a release build for me.

The extension reports as being supported by the loader, but the function addresses are null, so I get a segfault when it tries to call them.

I suspect this has something to do with the fact that the framework's instance constructor only enables VK_EXT_debug_utils if "defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)". I'm guessing that evaluates false in my release build. See framework/core/instance.cpp lines 137-153.

SaschaWillems commented:
@gary-sweet: Good catch. The extension is only enabled if one of those defines is active, and the sample doesn't properly check for it. I'll try to catch this in the sample and show a warning if the sample is run without the extension enabled.

SaschaWillems committed:
Check against actually enabled instance extension to determine if debug utils can be used

Display a clear warning with reasons if debug utilities are not available

SaschaWillems commented:
The sample will now check if the extension was actually enabled at instance creation time (instead of just checking if it's supported). If that's not the case, it'll also display a clear warning as to why debug utilities are not available. This worked for me in all combinations, including release with validation layer define active and without.

gary-sweet reviewed (commented):
The crashes are fixed for me now, thanks.

I don't really have any mechanism to test that the debug actually work however, so I'll leave it to those that do to approve.

JoseEmilio-ARM commented on README.md:
```diff
@@ -77,6 +77,8 @@ Additionally you may find the following links useful:
 - **AFBC**
   - [Appropriate use of AFBC](./samples/performance/afbc/afbc_tutorial.md)
+- **Debug utilities**
+  - [Using debug utilities for advanced debugging](./samples/extensions/debug_utils/debug_utils_tutorial.md)
```
Would you also like to add a link to your previous tutorial https://github.com/KhronosGroup/Vulkan-Samples/blob/master/samples/api/texture_mipmap_generation/texture_mipmap_generation_tutorial.md ?

Also, should we consider splitting these up into API / Performance / Extensions sections?

SaschaWillems replied:
Sure, forgot to do that when adding that sample. Will do that in another PR though.

Splitting may make sense once the list grows.

SaschaWillems replied:
I have moved this to a new issue, see #78.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
````diff
@@ -0,0 +1,255 @@
+
+## Setup
+
+**Note**: Enabling the extension is done inside the framework, see the ```Instance``` class in ```instance.cpp``` for details.
````
Could we add a link to the file here? https://github.com/KhronosGroup/Vulkan-Samples/blob/master/framework/core/instance.cpp

SaschaWillems replied:
Sure, I'll put up a relative link to that cpp file in there.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
````diff
@@ -0,0 +1,255 @@
+
+**Note**: Enabling the extension is done inside the framework, see the ```Instance``` class in ```instance.cpp``` for details.
+
+Enabling the functionality for the debug utilities is done by adding the extension to the list of extension to enable at instance level. 
````
"list of extension**s**"

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
````diff
@@ -0,0 +1,255 @@
+
+After creating your instance with the ```VK_EXT_debug_utils``` extension enabled, you'll be able to use the debug functions that have been introduced with it.
+
+**Note:**: Depending on the implementation and loader you're using you may need to manually get the function pointers for these via ```vkGetInstanceProcAddr``` before you can use these. 
````
double ':'

ZandroFargnoli commented on samples/extensions/debug_utils/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+get_filename_component(PARENT_DIR ${CMAKE_CURRENT_LIST_DIR} PATH)
+get_filename_component(CATEGORY_NAME ${PARENT_DIR} NAME)
+
+add_sample_with_tags(
```
If youre not using tags you should be able to just use `add_sample`

JoseEmilio-ARM replied:
@SaschaWillems did you try using `add_sample`?

SaschaWillems replied:
Yes, already changed it locally. Will push later today.

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

# Conflicts:
#	README.md
#	framework/core/instance.cpp

SaschaWillems committed:
Updated debug utils creation to use pNext instance structure chaining

SaschaWillems commented:
I have rebased my changes to reflect the recently merged PRs. The debug utils creation info is now passed via to the instance via pNext chaining.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
```diff
@@ -0,0 +1,255 @@
+
+<a href="./images/renderdoc_no_labels.jpg"><img src="./images/renderdoc_no_labels.jpg" style="vertical-align:top; border:1px solid black" height="400px"></a> <a href="./images/renderdoc_with_labels.jpg"><img src="./images/renderdoc_with_labels.jpg" style="border:1px solid black" height="400px"></a>
+
+Color choice aside, it's clear that the event tree with added debug labels is much easier to navigate- 
```
replace '-' with '.'?

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
````diff
@@ -0,0 +1,255 @@
+
+Once you start a new label via ```vkCmd/QueueBeginDebugUtilsLabelEXT``` all commands submitted to that command buffer or queue are encapsulated by that label until you end it via ```vkCmd/QueueEndDebugUtilsLabelEXT``` whereas a call to ```vkCmd/QueueInsertDebugUtilsLabelEXT``` simply inserts a marker at the current command buffer or queue command position;
+
+For convenience, the sample wraps those functions into dedicated functions.
````
would it be possible to move these wrappers to a separate file in the framework for other samples to use too?

SaschaWillems replied:
Yes, I think these would make sense in the framework too. For the sake of completeness I'd still like to keep them in the source for this sample.

Maybe move this to a separate issue?

SaschaWillems replied:
See #77

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
```diff
@@ -0,0 +1,255 @@
+
+Finding "Shader Module 257" or any of the "Graphics Pipeline 259/260" in your code could will prove tricky to impossible.
+
+But if you're using the new extension to add meaningful names to your, connecting the resources and finding them in your application becomes straightforward:
```
"to add meaningful names to your Vulkan objects"?

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
```diff
@@ -0,0 +1,255 @@
+
+But if you're using the new extension to add meaningful names to your, connecting the resources and finding them in your application becomes straightforward:
+
+<a href="./images/renderdoc_resource_inspector_names.jpg"><img src="./images/renderdoc_resource_inspector_names.jpg" style="border:1px solid black"></a>
```
what do the numbers in the image mean? (similarly for renderdoc_pipeline_state_names.jpg)

SaschaWillems replied:
Those number are generated by RenderDoc. I'll add this information to the tutorial.

JoseEmilio-ARM replied:
oh, I meant the numbers in the red circles that were added on top of the images

SaschaWillems replied:
You're right, missed that one. Will add hints to at what those numbers refer to.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils_tutorial.md:
````diff
@@ -0,0 +1,255 @@
+
+This tutorial, along with the accompanying example code, demonstrates the use of the ```VK_EXT_debug_utils``` extension to setup a validation layer messenger callback and pass additional debugging information to debuggers like [RenderDoc](https://renderdoc.org/).
+
+```VK_EXT_debug_utils``` has been introcuded based on feedback for the initial Vulkan debugging extensions ```VK_EXT_debug_report``` and ```VK_EXT_debug_marker```, combining these into a single instance extensions with some added functionality.
````
could we also add a link to the spec for these?

SaschaWillems replied:
Done. Though the spec link contains a version number (and sadly takes ages to load), so the link has to be updated accordingly.

JoseEmilio-ARM reviewed (changes requested):
Excellent sample, thank you!

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+
+		dependencies[0].srcSubpass      = VK_SUBPASS_EXTERNAL;
+		dependencies[0].dstSubpass      = 0;
+		dependencies[0].srcStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
```
VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT?

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		dependencies[0].dstStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[0].srcAccessMask   = VK_ACCESS_MEMORY_READ_BIT;
+		dependencies[0].dstAccessMask   = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[0].dependencyFlags = VK_DEPENDENCY_BY_REGION_BIT;
```
Not a framebuffer-local dependency. Remove?

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		dependencies[1].srcSubpass      = 0;
+		dependencies[1].dstSubpass      = VK_SUBPASS_EXTERNAL;
+		dependencies[1].srcStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[1].dstStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
```
The intention is to read the output of this pass in the fragment shader in the following pass, so `VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT`?

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		dependencies[1].dstStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
+		dependencies[1].srcAccessMask   = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[1].dstAccessMask   = VK_ACCESS_MEMORY_READ_BIT;
+		dependencies[1].dependencyFlags = VK_DEPENDENCY_BY_REGION_BIT;
```
Remove `VK_DEPENDENCY_BY_REGION_BIT`.

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+
+		dependencies[0].srcSubpass      = VK_SUBPASS_EXTERNAL;
+		dependencies[0].dstSubpass      = 0;
+		dependencies[0].srcStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
```
`VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT`?

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		dependencies[0].dstStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[0].srcAccessMask   = VK_ACCESS_MEMORY_READ_BIT;
+		dependencies[0].dstAccessMask   = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[0].dependencyFlags = VK_DEPENDENCY_BY_REGION_BIT;
```
As above, remove `VK_DEPENDENCY_BY_REGION_BIT`.

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		dependencies[1].srcStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[1].dstStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
+		dependencies[1].srcAccessMask   = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[1].dstAccessMask   = VK_ACCESS_MEMORY_READ_BIT;
```
Any non-zero access-mask doesn't make sense for `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT` -- remove.

cforfang replied:
Correction, the output of this pass feeds into the 3rd pass doesn't it? So `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT` -> `VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT`, but keep `VK_ACCESS_MEMORY_READ_BIT`?

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		dependencies[1].dstStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
+		dependencies[1].srcAccessMask   = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[1].dstAccessMask   = VK_ACCESS_MEMORY_READ_BIT;
+		dependencies[1].dependencyFlags = VK_DEPENDENCY_BY_REGION_BIT;
```
Remove; not a framebuffer-local dependency.

cforfang commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1135 @@
+		}
+
+		/*
+			Note: Explicit synchronization is not required between the render pass, as this is done implicit via sub pass dependencies
```
Consider something as simple as;
> Note: Synchronization between render passes is handled via sub pass dependencies.

This is because taking about 'explicit' and 'implicit' synchronization could be confusing; the synchronization is not really implicit here -- it's just defined elsewhere. Additionally, talking about 'implicit' sub pass dependencies risks confusing the current explicit ones with the the (actually) implicit subpass dependency which will be added by the implementation when custom VK_SUBPASS_EXTERNAL dependencies are not present.

SaschaWillems committed:
Updated readme

Fix spelling, add links

SaschaWillems committed:
Subpass dependencies

SaschaWillems commented:
@JoseEmilio-ARM: I have fixed the types, and added some links.

@cforfang: I have changed the subpass dependencies. I noticed that those are probably wrong in several of the samples, so you may want to check these too and if possible provide a PR to fix them.

SaschaWillems committed:
Clang format

SaschaWillems committed:
Added hints to image markings

SaschaWillems committed:
Expanded setup chapter with differences between implicit and explicit debug utilities messenger creation

SaschaWillems committed:
Explicitly create callback, update readme

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1131 @@
+		dependencies[0].dstSubpass    = 0;
+		dependencies[0].srcStageMask  = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[0].dstStageMask  = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[0].srcAccessMask = VK_ACCESS_MEMORY_READ_BIT;
```
one final suggestion would be to use `VK_ACCESS_COLOR_ATTACHMENT_READ_BIT` here to be a little bit more specific and enable driver optimizations (e.g. subpass fusion)

SaschaWillems replied:
Will be fixed with the next commit.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1131 @@
+		subpass.pColorAttachments    = color_references.data();
+		subpass.colorAttachmentCount = 1;
+
+		// Use subpass dependencies for attachment layput transitions
```
layout

SaschaWillems replied:
Thanks for noticing. Will be fixed.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1131 @@
+		dependencies[0].dstSubpass    = 0;
+		dependencies[0].srcStageMask  = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[0].dstStageMask  = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[0].srcAccessMask = VK_ACCESS_MEMORY_READ_BIT;
```
`srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT`

SaschaWillems replied:
Will be fixed with the next commit.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1131 @@
+		dependencies[1].srcStageMask  = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[1].dstStageMask  = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+		dependencies[1].srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[1].dstAccessMask = VK_ACCESS_MEMORY_READ_BIT;
```
`dstAccessMask = VK_ACCESS_SHADER_READ_BIT`

SaschaWillems replied:
Will be fixed with the next commit.

JoseEmilio-ARM commented on samples/extensions/debug_utils/debug_utils.cpp:
```diff
@@ -0,0 +1,1131 @@
+		dependencies[1].srcStageMask  = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+		dependencies[1].dstStageMask  = VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+		dependencies[1].srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+		dependencies[1].dstAccessMask = VK_ACCESS_MEMORY_READ_BIT;
```
`dstAccessMask = VK_ACCESS_SHADER_READ_BIT`

SaschaWillems replied:
Will be fixed with the next commit.

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

# Conflicts:
#	README.md
#	framework/core/instance.cpp

SaschaWillems committed:
Add debug utilities sample to api readme

SaschaWillems committed:
Use add_sample for CMake

SaschaWillems committed:
Fixed typo

SaschaWillems committed:
Properly initialize sub pass dependencies

SaschaWillems committed:
Adjusted subpass dependency access masks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=56) <br/>All committers have signed the CLA.

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

# Conflicts:
#	framework/core/instance.cpp

SaschaWillems committed:
Manual Fixes to accommodate for recent framework changes on master

mattruschnv reviewed (approved):
I get some warnings, but it looks like they're harmless. The sample runs fine for me on my RTX 2700.

[warning] (Swapchain) Image extent (0, 0) not supported. Selecting (1280, 720).
[warning] (Swapchain) Surface format (VK_FORMAT_UNDEFINED, VK_COLORSPACE_SRGB_NONLINEAR_KHR) not supported. Selecting (VK_FORMAT_B8G8R8A8_SRGB, VK_COLORSPACE_SRGB_NONLINEAR_KHR).
[warning] (Swapchain) Composite alpha 'VK_COMPOSITE_ALPHA_INHERIT_BIT_KHR' not supported. Selecting 'VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR.
